elasticdl/python/master/servicer.py

```diff
@@ -401,16 +399,19 @@ def ReportGradient(self, request, _):
                     self._gradient_sum[k] = self._gradient_sum[k] + v
                 else:
                     self._gradient_sum[k] = v
-
             self._grad_n += 1
-            # staleness-aware learning rate modulation
-            if self._lr_modulation:
-                staleness = max(1, self._version - request.model_version)
-                self._lr_modulation.set_multiplier(1.0 / staleness)
-            if self._use_async or self._grad_n >= self._grad_to_wait:
-                self._update_model()
-                self._update_evaluation()
-                self._update_checkpoint()
+        else:
+            # TODO: do not accumulate gradients but apply directly.
+            pass
+
+        # staleness-aware learning rate modulation
+        if self._lr_modulation:
+            staleness = max(1, self._version - request.model_version)
+            self._lr_modulation.set_multiplier(1.0 / staleness)
+        if self._use_async or self._grad_n >= self._grad_to_wait:
+            self._update_model()
+        if not self._use_async:
```

Comment:
`lock.acquire` and `lock.release` appears in multiple places. It is easy to mess things up.

Reply:
One possible idea is to move some intermediate code outside to another func to make the acquire->func->release clearer.

Reply:
changed